autoblocks/_impl/testing/run.py

```diff
@@ -0,0 +1,301 @@
+            testCaseHash=test_case_hash,
+            evaluatorExternalId=evaluator_id,
+            error=dict(
+                name=type(error).__name__,
+                message=str(error),
+                stacktrace=traceback.format_exc(),
+            ),
+        ),
+    )
+
+
+async def gather_with_max_concurrency(
+    max_concurrency: int,
+    *coroutines: Coroutine,
+) -> None:
+    """
+    Borrowed from https://stackoverflow.com/a/61478547
+    """
+    semaphore = asyncio.Semaphore(max_concurrency)
+
+    async def sem_coro(coro: Coroutine):
+        async with semaphore:
+            return await coro
+
+    # return_exceptions=True causes exceptions to be returned as values instead
+    # of propagating them to the caller. this is similar in behavior to Promise.allSettled
+    await asyncio.gather(*(sem_coro(c) for c in coroutines), return_exceptions=True)
+
+
+async def evaluate_output(
+    test_id: str,
+    test_case: BaseTestCase,
+    output: Any,
+    evaluator: BaseEvaluator,
+):
+    evaluation: Optional[Evaluation] = None
+
+    if inspect.iscoroutinefunction(evaluator.evaluate):
+        try:
+            evaluation = await evaluator.evaluate(test_case, output)
+        except Exception as err:
+            await send_error(
+                test_id=test_id,
+                test_case_hash=test_case._cached_hash,
+                evaluator_id=evaluator.id,
+                error=err,
+            )
+    else:
+        try:
+            ctx = contextvars.copy_context()
+            evaluation = await loop.run_in_executor(
+                None,
+                ctx.run,
+                evaluator.evaluate,
+                test_case,
+                output,
+            )
+        except Exception as err:
+            await send_error(
+                test_id=test_id,
+                test_case_hash=test_case._cached_hash,
+                evaluator_id=evaluator.id,
+                error=err,
+            )
+
+    if evaluation is None:
+        return
+
+    await client.post(
+        "/evals",
+        json=dict(
+            testExternalId=test_id,
+            testCaseHash=test_case._cached_hash,
+            evaluatorExternalId=evaluator.id,
+            score=evaluation.score,
+            threshold=dataclasses.asdict(evaluation.threshold) if evaluation.threshold else None,
+        ),
+    )
+
+
+async def run_test_case(
+    test_id: str,
+    test_case: BaseTestCase,
+    evaluators: List[BaseEvaluator],
+    fn: Callable,
+    max_evaluator_concurrency: int,
+):
+    current_test_id_var.set(test_id)
+    current_test_case_var.set(test_case)
+
+    output = None
+
+    if inspect.iscoroutinefunction(fn):
+        try:
+            output = await fn(test_case)
+        except Exception as err:
+            await send_error(
+                test_id=test_id,
+                test_case_hash=test_case._cached_hash,
+                evaluator_id=None,
+                error=err,
+            )
+    else:
```

Comment:
These if else statements - essentially checking if `fn` was defined using the `async def` function syntax right?

Reply:
yeah exactly: https://docs.python.org/3/library/inspect.html#inspect.iscoroutinefunction